In [1]:
%%capture
import sys
!{sys.executable} -m pip install requests pandas geopy bs4 folium sklearn matplotlib

In [2]:
import folium
import requests
import pandas as pd
from geopy.geocoders import Nominatim
from bs4 import BeautifulSoup
import json
import numpy as np
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

In [3]:
# Setting credentials required by foursquare
CLIENT_ID = 'DEPKYDHCHC40DUDYBECM3Y1GYBGOFGRTODGR4QLNQ5VPNLBU'
CLIENT_SECRET = 'FWKDHU4URB1WHQI5IKQ5QN4RMLEWOASA34AHLW3YNA4KCGH5'
VERSION = '20190808'

<h4>Scrape list of areas in Pune city from "makemytrip" website</h4>

In [4]:
# Function to scrape data from website and convert it into dataframe
geolocator = Nominatim(user_agent="coursera agent")
def create_areas_data_frame(): 
    with open('area_index.html') as f:
        soup = BeautifulSoup(f, parser='lxml')

    city_areas = soup.find_all('div', attrs={'class': 'city-listing'})
    areas_dict = {'Area': [], 'Latitude': [], 'Longitude': []}
    for areas in city_areas:
        area_cat = areas.find('p', attrs={'class': 'key-values'})
        for area_parsed in area_cat.find_all('a'):
            area_short = area_parsed.text.split(',')[0]
            area = '{}, Pune'.format(area_short)
            try:
                location = geolocator.geocode(area)
                latitude = location.latitude
                longitude = location.longitude
            except Exception as e:
                pass
            else:
                areas_dict['Area'].append(area_short)
                areas_dict['Latitude'].append(latitude)
                areas_dict['Longitude'].append(longitude)

    df_areas = pd.DataFrame(areas_dict)
    # df_areas.to_csv('pune_areas.csv')
    return df_areas

Load data in a dataframe and discard areas which lie in the outskirts of the city

In [5]:
# Load scraped data from csv
df_pune = pd.read_csv('pune_areas.csv')
df_pune.drop(['Unnamed: 0', 'Unnamed: 0.1'], axis=1, inplace=True)
df_pune = df_pune[(df_pune['Latitude'] < 18.6) & (df_pune['Latitude'] > 18.35)
                  & (df_pune['Longitude'] < 74.0) & (df_pune['Longitude'] > 73.6)]
print(df_pune.shape)
df_pune.head(10)

(69, 3)


Area   Latitude  Longitude
0              Aundh  18.561883  73.810196
3        Anand Nagar  18.478490  73.821326
5           Balewadi  18.582027  73.768983
6              Baner  18.564243  73.776857
7   Bund Garden Road  18.533404  73.878625
10      Budhwar Peth  18.517560  73.858037
11             Baner  18.564243  73.776857
12      Central Pune  18.520306  73.854319
17  Dhole Patil Road  18.534908  73.876258
18   Deccan Gymkhana  18.515893  73.841160

In total, there are 69 areas in Pune. The cell above shows dataframe with areas and their latitudes and longitudes

In [6]:
location = geolocator.geocode('Pune')
pune_map = folium.Map(location=[location.latitude, location.longitude], zoom_start=11)
for lat, lon, area in zip(df_pune['Latitude'], df_pune['Longitude'], df_pune['Area']):
    label = folium.Popup(area)
    folium.CircleMarker(
        [lat, lon],
        color='blue',
        popup=label,
        radius=8,
        fill_color='yellow',
        fill=True,
        fill_opacity=0.8
    ).add_to(pune_map)
pune_map

In [7]:
LIMIT=100
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&query=restaurant'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

<h4>Get list of restaurants under each area</h4>

In [8]:
pune_venues = getNearbyVenues(names=df_pune['Area'],
                                   latitudes=df_pune['Latitude'],
                                   longitudes=df_pune['Longitude']
                                  )

Aundh
Anand Nagar
Balewadi
Baner
Bund Garden Road
Budhwar Peth
Baner
Central Pune
Dhole Patil Road
Deccan Gymkhana
Dapodi
Fergusson College Road
Fursungi
Fatima Nagar
Ghorpadi
Gultekdi
Hadapsar
Hinjewadi
Hinjewadi
Kothrud
Kalyani Nagar
Karve Nagar
Koregaon Park
Kharadi
Katraj
Kondhwa
Khadki
Laxmi Road
Magarpatta
Mundhwa
Mangalwar Peth
MG Road
Mumbai - Bangalore Highway
Mundhwa - Kharadi Road
Municipal Colony
Nagar Road
Nagar - Pune Highway
New Airport Road
Nana Peth
Narayan Peth
Narhe
NIA Campus Road
Pirangut
Pune - Solapur Highway
Pimple Saudagar
Pune Central City
Pune
Ragvilas Society
Revenue Colony
Senapati Bapat Road
Shaniwar Peth
Somwar Peth
Sadashiv Peth
Sinhagad Road
Swargate
Sangamwadi Road
Shukrawar Peth
Sky Lounge
Saswad Road
Swargate
University Road
Viman Nagar
Vitthalrao Navale Rd
Vishrantwadi
Wadgaon Sheri
Wanowrie
Ward No. 8
Wagholi
Yerwada


In [9]:
print(pune_venues.shape)
pune_venues.head()

(623, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0        Aundh              18.561883               73.810196   
1        Aundh              18.561883               73.810196   
2        Aundh              18.561883               73.810196   
3        Aundh              18.561883               73.810196   
4        Aundh              18.561883               73.810196   

                    Venue  Venue Latitude  Venue Longitude  \
0  Picantos Mexican Grill       18.560654        73.812447   
1             Venkys Xprs       18.560550        73.808964   
2              Café Maroo       18.564801        73.809141   
3                The Loft       18.562714        73.807630   
4        Hyderabadi Spice       18.560594        73.810108   

         Venue Category  
0    Mexican Restaurant  
1  Fast Food Restaurant  
2     Korean Restaurant  
3     Indian Restaurant  
4     Indian Restaurant

API lists down 623 total restaurants in the city of Pune.

In [10]:
# Group by neighborhood/area
pune_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                             
Anand Nagar                           9                       9      9   
Aundh                                22                      22     22   
Baner                                28                      28     28   
Budhwar Peth                          5                       5      5   
Bund Garden Road                     27                      27     27   
...                                 ...                     ...    ...   
Wadgaon Sheri                         2                       2      2   
Wagholi                               5                       5      5   
Wanowrie                             13                      13     13   
Ward No. 8                            6                       6      6   
Yerwada                               2                       2      2   

                  Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                       
Anand Nagar                    9                9               9  
Aundh                         22               22              22  
Baner                         28               28              28  
Budhwar Peth                   5                5               5  
Bund Garden Road              27               27              27  
...                          ...              ...             ...  
Wadgaon Sheri                  2                2               2  
Wagholi                        5                5               5  
Wanowrie                      13               13              13  
Ward No. 8                     6                6               6  
Yerwada                        2                2               2  

[61 rows x 6 columns]

In [11]:
print('There are {} uniques categories.'.format(len(pune_venues['Venue Category'].unique())))

There are 50 uniques categories.


In [12]:
# one hot encoding
pune_onehot = pd.get_dummies(pune_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
pune_onehot['Neighborhood'] = pune_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [pune_onehot.columns[-1]] + list(pune_onehot.columns[:-1])
pune_onehot = pune_onehot[fixed_columns]

pune_onehot.head()

Neighborhood  American Restaurant  Asian Restaurant  BBQ Joint  Bakery  \
0        Aundh                    0                 0          0       0   
1        Aundh                    0                 0          0       0   
2        Aundh                    0                 0          0       0   
3        Aundh                    0                 0          0       0   
4        Aundh                    0                 0          0       0   

   Bistro  Breakfast Spot  Buffet  Burger Joint  Burrito Place  ...  \
0       0               0       0             0              0  ...   
1       0               0       0             0              0  ...   
2       0               0       0             0              0  ...   
3       0               0       0             0              0  ...   
4       0               0       0             0              0  ...   

   Pizza Place  Punjabi Restaurant  Restaurant  Sandwich Place  \
0            0                   0           0               0   
1            0                   0           0               0   
2            0                   0           0               0   
3            0                   0           0               0   
4            0                   0           0               0   

   Seafood Restaurant  Snack Place  South Indian Restaurant  Steakhouse  \
0                   0            0                        0           0   
1                   0            0                        0           0   
2                   0            0                        0           0   
3                   0            0                        0           0   
4                   0            0                        0           0   

   Thai Restaurant  Vegetarian / Vegan Restaurant  
0                0                              0  
1                0                              0  
2                0                              0  
3                0                              0  
4                0                              0  

[5 rows x 51 columns]

In [13]:
pune_onehot.shape

(623, 51)

In [14]:
pune_grouped = pune_onehot.groupby('Neighborhood').mean().reset_index()
pune_grouped

Neighborhood  American Restaurant  Asian Restaurant  BBQ Joint  \
0        Anand Nagar                  0.0               0.0   0.000000   
1              Aundh                  0.0               0.0   0.000000   
2              Baner                  0.0               0.0   0.000000   
3       Budhwar Peth                  0.0               0.0   0.000000   
4   Bund Garden Road                  0.0               0.0   0.037037   
..               ...                  ...               ...        ...   
56     Wadgaon Sheri                  0.0               0.0   0.000000   
57           Wagholi                  0.0               0.0   0.000000   
58          Wanowrie                  0.0               0.0   0.000000   
59        Ward No. 8                  0.0               0.0   0.000000   
60           Yerwada                  0.0               0.0   0.000000   

      Bakery    Bistro  Breakfast Spot    Buffet  Burger Joint  Burrito Place  \
0   0.333333  0.000000        0.000000  0.000000           0.0       0.000000   
1   0.000000  0.000000        0.090909  0.000000           0.0       0.000000   
2   0.000000  0.000000        0.071429  0.000000           0.0       0.000000   
3   0.200000  0.000000        0.000000  0.000000           0.0       0.000000   
4   0.074074  0.037037        0.037037  0.037037           0.0       0.000000   
..       ...       ...             ...       ...           ...            ...   
56  0.000000  0.000000        0.000000  0.000000           0.0       0.000000   
57  0.000000  0.000000        0.000000  0.000000           0.0       0.000000   
58  0.153846  0.000000        0.076923  0.000000           0.0       0.000000   
59  0.166667  0.000000        0.000000  0.000000           0.0       0.166667   
60  0.000000  0.000000        0.000000  0.000000           0.0       0.000000   

    ...  Pizza Place  Punjabi Restaurant  Restaurant  Sandwich Place  \
0   ...     0.111111                 0.0    0.000000        0.000000   
1   ...     0.000000                 0.0    0.090909        0.000000   
2   ...     0.000000                 0.0    0.000000        0.000000   
3   ...     0.000000                 0.0    0.000000        0.000000   
4   ...     0.074074                 0.0    0.037037        0.000000   
..  ...          ...                 ...         ...             ...   
56  ...     0.000000                 0.0    0.000000        0.000000   
57  ...     0.200000                 0.0    0.200000        0.000000   
58  ...     0.000000                 0.0    0.000000        0.076923   
59  ...     0.000000                 0.0    0.000000        0.000000   
60  ...     0.000000                 0.0    0.000000        0.000000   

    Seafood Restaurant  Snack Place  South Indian Restaurant  Steakhouse  \
0             0.000000     0.111111                 0.000000         0.0   
1             0.000000     0.045455                 0.000000         0.0   
2             0.142857     0.142857                 0.071429         0.0   
3             0.000000     0.000000                 0.000000         0.0   
4             0.000000     0.000000                 0.000000         0.0   
..                 ...          ...                      ...         ...   
56            0.000000     0.000000                 0.500000         0.0   
57            0.000000     0.000000                 0.000000         0.0   
58            0.000000     0.076923                 0.000000         0.0   
59            0.000000     0.333333                 0.000000         0.0   
60            0.000000     0.500000                 0.000000         0.0   

    Thai Restaurant  Vegetarian / Vegan Restaurant  
0               0.0                       0.000000  
1               0.0                       0.045455  
2               0.0                       0.000000  
3               0.0                       0.000000  
4               0.0                       0.000000  
..              ...                            ...

In [15]:
pune_grouped.shape

(61, 51)

Listing down top five types of restaurants in each area

In [16]:
num_top_venues = 5

for hood in pune_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = pune_grouped[pune_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Anand Nagar----
                  venue  freq
0                Bakery  0.33
1  Fast Food Restaurant  0.22
2     Indian Restaurant  0.11
3           Snack Place  0.11
4                  Café  0.11


----Aundh----
                  venue  freq
0     Indian Restaurant  0.27
1  Fast Food Restaurant  0.23
2            Restaurant  0.09
3        Breakfast Spot  0.09
4            Food Court  0.09


----Baner----
                     venue  freq
0        Indian Restaurant  0.29
1     Fast Food Restaurant  0.21
2              Snack Place  0.14
3       Seafood Restaurant  0.14
4  South Indian Restaurant  0.07


----Budhwar Peth----
                     venue  freq
0        Indian Restaurant   0.6
1                   Bakery   0.2
2        Indian Sweet Shop   0.2
3  North Indian Restaurant   0.0
4         Kebab Restaurant   0.0


----Bund Garden Road----
                  venue  freq
0     Indian Restaurant  0.33
1    Chinese Restaurant  0.11
2                  Café  0.07
3                Baker

In [17]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create area-wise top 10 most common type of restaurants/food outlets

In [18]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = pune_grouped['Neighborhood']

for ind in np.arange(pune_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(pune_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0       Anand Nagar                Bakery  Fast Food Restaurant   
1             Aundh     Indian Restaurant  Fast Food Restaurant   
2             Baner     Indian Restaurant  Fast Food Restaurant   
3      Budhwar Peth     Indian Restaurant     Indian Sweet Shop   
4  Bund Garden Road     Indian Restaurant    Chinese Restaurant   

  3rd Most Common Venue          4th Most Common Venue  \
0                  Café              Indian Restaurant   
1            Food Court                 Breakfast Spot   
2           Snack Place             Seafood Restaurant   
3                Bakery  Vegetarian / Vegan Restaurant   
4                  Café                    Pizza Place   

     5th Most Common Venue 6th Most Common Venue    7th Most Common Venue  \
0              Snack Place           Pizza Place  Comfort Food Restaurant   
1               Restaurant     Korean Restaurant       Mexican Restaurant   
2  South Indian Restaurant        Breakfast Spot                     Café   
3                 Creperie            Food Truck               Food Court   
4                   Bakery  Fast Food Restaurant         Kebab Restaurant   

           8th Most Common Venue          9th Most Common Venue  \
0                     Food Court                           Food   
1      Middle Eastern Restaurant  Vegetarian / Vegan Restaurant   
2  Vegetarian / Vegan Restaurant        Comfort Food Restaurant   
3                           Food           Fast Food Restaurant   
4             Italian Restaurant                         Buffet   

  10th Most Common Venue  
0     Falafel Restaurant  
1            Snack Place  
2             Food Court  
3     Falafel Restaurant  
4             Restaurant

In [ ]:
Use k-means clustering method to cluster city in 3 segments

In [19]:
# set number of clusters
kclusters = 3

pune_grouped_clustering = pune_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(pune_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 1, 1, 4, 1, 2, 0, 1, 1, 1], dtype=int32)

In [20]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

pune_merged = df_pune

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
pune_merged = pune_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Area')

pune_merged.head() # check the last columns!

Area   Latitude  Longitude  Cluster Labels  \
0             Aundh  18.561883  73.810196             1.0   
3       Anand Nagar  18.478490  73.821326             0.0   
5          Balewadi  18.582027  73.768983             NaN   
6             Baner  18.564243  73.776857             1.0   
7  Bund Garden Road  18.533404  73.878625             1.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0     Indian Restaurant  Fast Food Restaurant            Food Court   
3                Bakery  Fast Food Restaurant                  Café   
5                   NaN                   NaN                   NaN   
6     Indian Restaurant  Fast Food Restaurant           Snack Place   
7     Indian Restaurant    Chinese Restaurant                  Café   

  4th Most Common Venue    5th Most Common Venue 6th Most Common Venue  \
0        Breakfast Spot               Restaurant     Korean Restaurant   
3     Indian Restaurant              Snack Place           Pizza Place   
5                   NaN                      NaN                   NaN   
6    Seafood Restaurant  South Indian Restaurant        Breakfast Spot   
7           Pizza Place                   Bakery  Fast Food Restaurant   

     7th Most Common Venue          8th Most Common Venue  \
0       Mexican Restaurant      Middle Eastern Restaurant   
3  Comfort Food Restaurant                     Food Court   
5                      NaN                            NaN   
6                     Café  Vegetarian / Vegan Restaurant   
7         Kebab Restaurant             Italian Restaurant   

           9th Most Common Venue 10th Most Common Venue  
0  Vegetarian / Vegan Restaurant            Snack Place  
3                           Food     Falafel Restaurant  
5                            NaN                    NaN  
6        Comfort Food Restaurant             Food Court  
7                         Buffet             Restaurant

In [21]:
pune_merged.dropna(axis=0, inplace=True)

In [22]:
pune_merged['Cluster Labels'] = pune_merged['Cluster Labels'].astype(int)
pune_merged.head(20)

Area   Latitude  Longitude  Cluster Labels  \
0                    Aundh  18.561883  73.810196               1   
3              Anand Nagar  18.478490  73.821326               0   
6                    Baner  18.564243  73.776857               1   
7         Bund Garden Road  18.533404  73.878625               1   
10            Budhwar Peth  18.517560  73.858037               4   
11                   Baner  18.564243  73.776857               1   
12            Central Pune  18.520306  73.854319               2   
17        Dhole Patil Road  18.534908  73.876258               1   
18         Deccan Gymkhana  18.515893  73.841160               1   
21                  Dapodi  18.580846  73.832775               0   
23  Fergusson College Road  18.524148  73.838591               1   
25            Fatima Nagar  18.504510  73.902129               1   
27                Ghorpadi  18.526854  73.900096               4   
28                Gultekdi  18.493894  73.867633               1   
30               Hinjewadi  18.592094  73.757142               1   
31               Hinjewadi  18.592094  73.757142               1   
32                 Kothrud  18.503889  73.807673               4   
33           Kalyani Nagar  18.548138  73.902551               1   
34             Karve Nagar  18.489404  73.821324               3   
35           Koregaon Park  18.537553  73.893925               0   

   1st Most Common Venue    2nd Most Common Venue  \
0      Indian Restaurant     Fast Food Restaurant   
3                 Bakery     Fast Food Restaurant   
6      Indian Restaurant     Fast Food Restaurant   
7      Indian Restaurant       Chinese Restaurant   
10     Indian Restaurant        Indian Sweet Shop   
11     Indian Restaurant     Fast Food Restaurant   
12     Indian Restaurant              Snack Place   
17     Indian Restaurant       Chinese Restaurant   
18     Indian Restaurant                     Café   
21            Food Court           Breakfast Spot   
23     Indian Restaurant             Burger Joint   
25     Indian Restaurant           Breakfast Spot   
27     Indian Restaurant  Comfort Food Restaurant   
28        Breakfast Spot             Burger Joint   
30            Restaurant        Indian Restaurant   
31            Restaurant        Indian Restaurant   
32     Indian Restaurant      American Restaurant   
33     Indian Restaurant                     Café   
34            Food Truck              Snack Place   
35                Bakery         Asian Restaurant   

            3rd Most Common Venue          4th Most Common Venue  \
0                      Food Court                 Breakfast Spot   
3                            Café              Indian Restaurant   
6                     Snack Place             Seafood Restaurant   
7                            Café                    Pizza Place   
10                         Bakery  Vegetarian / Vegan Restaurant   
11                    Snack Place             Seafood Restaurant   
12                         Bakery                  Burrito Place   
17                         Bakery                           Café   
18  Vegetarian / Vegan Restaurant                    Snack Place   
21                     Restaurant                      Cafeteria   
23               Asian Restaurant                 Sandwich Place   
25                    Pizza Place                           Café   
27                      Gastropub                     Food Truck   
28  Vegetarian / Vegan Restaurant                       Creperie   
30                 Breakfast Spot       Mediterranean Restaurant   
31                 Breakfast Spot       Mediterranean Restaurant   
32                     Restaurant           Fast Food Restaurant   
33                    Pizza Place             Italian Restaurant   
34                         Bakery                 Breakfast Spot   
35              Indian Restaurant      Middle Eastern Restaurant   

            5th Most Common Venue          6th Mos

In [29]:
# create map
map_clusters = folium.Map(location=[location.latitude, location.longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(pune_merged['Latitude'], pune_merged['Longitude'], pune_merged['Area'], pune_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [24]:
# Cluster 0
pune_merged.loc[pune_merged['Cluster Labels'] == 0, pune_merged.columns[[1] + list(range(5, pune_merged.shape[1]))]]

Latitude          2nd Most Common Venue    3rd Most Common Venue  \
3    18.478490           Fast Food Restaurant                     Café   
21   18.580846                 Breakfast Spot               Restaurant   
35   18.537553               Asian Restaurant        Indian Restaurant   
40   18.477746  Vegetarian / Vegan Restaurant        Indian Restaurant   
56   18.557350                Thai Restaurant                BBQ Joint   
57   18.589521                    Snack Place     Fast Food Restaurant   
62   18.573425                           Café     Fast Food Restaurant   
79   18.525574                     Restaurant              Snack Place   
82   18.521766                           Café  Comfort Food Restaurant   
102  18.518231                           Café        Indian Restaurant   
106  18.488368                           Café                   Bakery   

         4th Most Common Venue          5th Most Common Venue  \
3            Indian Restaurant                    Snack Place   
21                   Cafeteria             Chinese Restaurant   
35   Middle Eastern Restaurant                     Restaurant   
40        Fast Food Restaurant                    Pizza Place   
56                      Bakery              Indian Restaurant   
57                  Restaurant  Vegetarian / Vegan Restaurant   
62              Breakfast Spot        Comfort Food Restaurant   
79              Sandwich Place                    Pizza Place   
82           Indian Restaurant             Seafood Restaurant   
102       Fast Food Restaurant        Comfort Food Restaurant   
106                Snack Place                 Sandwich Place   

             6th Most Common Venue    7th Most Common Venue  \
3                      Pizza Place  Comfort Food Restaurant   
21   Vegetarian / Vegan Restaurant                 Creperie   
35              Italian Restaurant     Fast Food Restaurant   
40                            Café         Asian Restaurant   
56                      Restaurant       Italian Restaurant   
57              Chinese Restaurant               Food Court   
62                      Food Truck               Food Court   
79   Vegetarian / Vegan Restaurant       Chinese Restaurant   
82                  Sandwich Place     Fast Food Restaurant   
102                       Creperie                Gastropub   
106                 Breakfast Spot       Falafel Restaurant   

    8th Most Common Venue 9th Most Common Venue         10th Most Common Venue  
3              Food Court                  Food             Falafel Restaurant  
21             Food Truck                  Food           Fast Food Restaurant  
35                   Café              Creperie  Vegetarian / Vegan Restaurant  
40                 Bakery             BBQ Joint                 Breakfast Spot  
56            Pizza Place            Food Truck                       Creperie  
57                   Food    Falafel Restaurant    Eastern European Restaurant  
62                   Food    Falafel Restaurant    Eastern European Restaurant  
79                   Food  Fast Food Restaurant             Falafel Restaurant  
82             Food Court                  Food             Falafel Restaurant  
102            Food Truck            Food Court                           Food  
106                 Diner    Chinese Restaurant                       Creperie

In [25]:
# Cluster 1
pune_merged.loc[pune_merged['Cluster Labels'] == 1, pune_merged.columns[[1] + list(range(5, pune_merged.shape[1]))]]

Latitude 2nd Most Common Venue          3rd Most Common Venue  \
0    18.561883  Fast Food Restaurant                     Food Court   
6    18.564243  Fast Food Restaurant                    Snack Place   
7    18.533404    Chinese Restaurant                           Café   
11   18.564243  Fast Food Restaurant                    Snack Place   
17   18.534908    Chinese Restaurant                         Bakery   
18   18.515893                  Café  Vegetarian / Vegan Restaurant   
23   18.524148          Burger Joint               Asian Restaurant   
25   18.504510        Breakfast Spot                    Pizza Place   
28   18.493894          Burger Joint  Vegetarian / Vegan Restaurant   
30   18.592094     Indian Restaurant                 Breakfast Spot   
31   18.592094     Indian Restaurant                 Breakfast Spot   
33   18.548138                  Café                    Pizza Place   
37   18.453679        Breakfast Spot                           Food   
42   18.508405                  Café                     Restaurant   
45   18.514894                  Café                      BBQ Joint   
48   18.533670     Indian Restaurant                 Breakfast Spot   
53   18.545907       Thai Restaurant              Indian Restaurant   
58   18.573073      Asian Restaurant           Fast Food Restaurant   
68   18.598155                  Café                     Restaurant   
83   18.510776        Breakfast Spot  Vegetarian / Vegan Restaurant   
87   18.541712     Indian Restaurant             Seafood Restaurant   
93   18.511762     Indian Restaurant               Asian Restaurant   
99   18.537493             Gastropub               Asian Restaurant   
103  18.572630           Pizza Place        South Indian Restaurant   
108  18.580630     Indian Restaurant                     Food Court   

           4th Most Common Venue           5th Most Common Venue  \
0                 Breakfast Spot                      Restaurant   
6             Seafood Restaurant         South Indian Restaurant   
7                    Pizza Place                          Bakery   
11            Seafood Restaurant         South Indian Restaurant   
17                          Café              Italian Restaurant   
18                   Snack Place            Fast Food Restaurant   
23                Sandwich Place                      Restaurant   
25                          Café   Vegetarian / Vegan Restaurant   
28                      Creperie                       Gastropub   
30      Mediterranean Restaurant   Vegetarian / Vegan Restaurant   
31      Mediterranean Restaurant   Vegetarian / Vegan Restaurant   
33            Italian Restaurant                  Breakfast Spot   
37                          Café               Indian Restaurant   
42                Breakfast Spot                     Snack Place   
45                      Creperie                       Gastropub   
48   Eastern European Restaurant   Vegetarian / Vegan Restaurant   
53                   Pizza Place   Vegetarian / Vegan Restaurant   
58                          Café                        Creperie   
68       North Indian Restaurant  Multicuisine Indian Restaurant   
83                   Snack Place                      Restaurant   
87            Italian Restaurant              Mughlai Restaurant   
93                    Food Truck                      Food Court   
99            Italian Restaurant                      Food Court   
103                  Snack Place                          Bakery   
108                   Restaurant                     Pizza Place   

             6th Most Common Venue      7th Most Common Venue  \
0                Korean Restaurant         Mexican Restaurant   
6                   Breakfast Spot                       Café   
7             Fast Food Restaurant           Kebab Restaurant   
11                  Breakfast Spot                       Café   
17                     Pizza Place       Fast Food Restaurant   


In [26]:
# Cluster 2
pune_merged.loc[pune_merged['Cluster Labels'] == 2, pune_merged.columns[[1] + list(range(5, pune_merged.shape[1]))]]

Latitude          2nd Most Common Venue          3rd Most Common Venue  \
12   18.520306                    Snack Place                         Bakery   
52   18.459712                    Snack Place                         Bakery   
54   18.508672  Vegetarian / Vegan Restaurant                    Snack Place   
60   18.515579              Indian Restaurant                         Bakery   
63   18.511282                           Café                 Breakfast Spot   
69   18.510289              Indian Restaurant                           Café   
72   18.520306                    Snack Place                         Bakery   
75   18.520306                    Snack Place                         Bakery   
76   18.520306                    Snack Place                         Bakery   
81   18.519284              Indian Restaurant                           Food   
88   18.511354  Vegetarian / Vegan Restaurant                    Snack Place   
101  18.520306                    Snack Place                         Bakery   
105  18.550441        South Indian Restaurant             Chinese Restaurant   
107  18.520306                    Snack Place                         Bakery   
109  18.550361                     Donut Shop  Vegetarian / Vegan Restaurant   

             4th Most Common Venue          5th Most Common Venue  \
12                   Burrito Place  Vegetarian / Vegan Restaurant   
52                  Breakfast Spot                           Café   
54                          Bakery                 Breakfast Spot   
60                            Café                 Breakfast Spot   
63         Comfort Food Restaurant                     Food Truck   
69   Vegetarian / Vegan Restaurant                         Bistro   
72                   Burrito Place  Vegetarian / Vegan Restaurant   
75                   Burrito Place  Vegetarian / Vegan Restaurant   
76                   Burrito Place  Vegetarian / Vegan Restaurant   
81              Seafood Restaurant                 Sandwich Place   
88                  Breakfast Spot                 Sandwich Place   
101                  Burrito Place  Vegetarian / Vegan Restaurant   
105                     Food Truck                     Food Court   
107                  Burrito Place  Vegetarian / Vegan Restaurant   
109             Chinese Restaurant                     Food Truck   

             6th Most Common Venue 7th Most Common Venue  \
12                        Creperie            Food Truck   
52   Vegetarian / Vegan Restaurant              Creperie   
54                            Café    Seafood Restaurant   
60                  Sandwich Place            Restaurant   
63                      Food Court                  Food   
69            Fast Food Restaurant    Chinese Restaurant   
72                        Creperie            Food Truck   
75                        Creperie            Food Truck   
76                        Creperie            Food Truck   
81                          Bakery         Burrito Place   
88               Indian Sweet Shop                  Café   
101                       Creperie            Food Truck   
105                           Food  Fast Food Restaurant   
107                       Creperie            Food Truck   
109                     Food Court                  Food   

        8th Most Common Venue        9th Most Common Venue  \
12                 Food Court                         Food   
52                 Food Truck                   Food Court   
54                 Food Truck                    BBQ Joint   
60   Maharashtrian Restaurant           Seafood Restaurant   
63       Fast Food Restaurant           Falafel Restaurant   
69                  Gastropub                   Donut Shop   
72                 Food Court                         Food   
75                 Food Court                         Food   
76                 Food Court                         Food   
81   Maharashtrian Restaurant          